### DISCUSSION TITLE
Thread-safe flag for methods

### DISCUSSION THREAD

**Discussion by AlwaysHC** [dotnet/csharplang#7826] [Language Ideas] [answered]
It could be useful if methods would contain a "flag" saying they are thread-safe or not.

**Comment by alexrp** [comment]
What would that flag mean? What would it do?

This sounds suspiciously like `MethodImplOptions.Synchronized` which is widely considered a design mistake.

**Reply by AlwaysHC** [reply]
For instance, when you use EF you need to remember that it is not thread safe by design and you cannot use the dbcontext inside a Task.WhenAll method.
A compiler warning would be really appreciated there.

**Reply by CyrusNajmabadi** [reply, marked as answer]
> A compiler warning would be really appreciated there.

You should recommend that to EF. They can write analyzers for this sort of scenario.

**Comment by huoyaoyuan** [comment]
Thread safety is not a yes-or-no flag.

Can different methods be invoked at the same time? Are they grouped?
What about read-write model? For example, many implementations would support concurrent read, but only when write is not being invoked.

**Comment by siegfriedpammer** [comment]
That's what custom attributes are for...

```csharp
using System;
public class C {
    [ThreadSafe(true)]
    public void M() {
    }
}

[AttributeUsage(AttributeTargets.Method)]
public class ThreadSafeAttribute : Attribute
{
    public bool IsThreadSafe { get; }

    public ThreadSafeAttribute(bool value) 
    {
        this.IsThreadSafe = value;
    }
}
```

And done. ;-)

However, I agree with all the people arguing against such a flag.
